### DISCUSSION TITLE
QE d3q runtime error

### DISCUSSION THREAD

**Discussion by pschell5** [anharmonic/d3q#9] [General]
Hi,

I built QE, specifically version 7.1, with d3q and associated programs. When I try to run d3q.x, however, I receive a runtime error which is:

 Attempting to use an MPI routine before initializing MPICH

It appears that the MPI Init statements have been commented out. Interestingly d3_q2r.x appears to run without error (although I haven't run it to figure out what it is doing).

Do you have any advice about what build of this code works? Specifically so that I can get anharmonic calculations, thermal conductivity, three-phonon scattering, etc, using QE calculations?

Thanks very much,
Patrick Schelling
Associate Prof.
Dept. of Physics
University of Central Florida

**Comment by paulatz** [comment]
Hello Patrick,
if the d3q and qe versions are consistent this should not happen, but it is a bit difficult to ensure. Can you check with this bundle, which has QE and D3Q shipped together ? https://drive.google.com/file/d/1HwHek9sXn1C_zF11iHqdzDn_XW137xNr/view?usp=drivesdk
hth

**Reply by jamie-arcc** [reply]
Hello,
I am a system administrator for the HPC system Dr. Schelling is trying to use this on.  I am getting different errors than before using the bundle you provided and it looks like it's related to cmake.  I am using cmake because the traditional autotools-type configuration script for QE is very broken.  I think the issue with the QE 7.1 release bundle we have been trying to use is that the cmake files are downloading the wrong commit of d3q from github (and I was not able to find a d3q release for QE 7.1 as the d3q docs seem to have some conflicting information and the only release in github was for QE 7.0). The commit it is downloading:
```
$ git log
commit 87685245bc5bc14f226fc18e5126305501b368e8 (grafted, HEAD -> recorded_HEAD)
Author: Ye Luo <yeluo@anl.gov>
Date:   Sat Jan 1 00:36:29 2022 -0600
```
Is there a commit of d3q that is known to work with QE 7.1?  If so, maybe there could be some communication between the d3q team and the QE team to make sure that things are staying synchronized.

It looks like, with the bundle you provided, the cmake scripts are still trying to download d3q from github rather than using the local copy, but other parts of the cmake config downright fail.
my cmake command:
`cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=${APP_DIR} -DQE_CPP=$(which mpicc) -DCMAKE_C_FLAGS="" -DCMAKE_C_FLAGS_RELEASE="${CFLAGS} -DNDEBUG"     -DCMAKE_AR=$(which xiar) -DCMAKE_LINKER=$(which xild) -DCMAKE_Fortran_FLAGS="" -DCMAKE_Fortran_FLAGS_RELEASE="${FFLAGS} -DNDEBUG" -DQE_ENABLE_BENCHMARK=ON -DQE_ENABLE_HDF5=ON -DQE_ENABLE_LIBXC=ON -DQE_ENABLE_MPI=ON -DQE_ENABLE_SCALAPACK=ON -DENABLE_SCALAPACK_MPI=ON -DQE_FFTW_VENDOR=Intel_DFTI -DQE_ENABLE_PLUGINS="d3q" .. `
yields:
[cmake.log](https://github.com/anharmonic/d3q/files/9499274/cmake.log)


Thanks,
Jamie

**Reply by paulatz** [reply]
Thank you for reporting, it is a bit of a mess, because cmake refuses to work when some sub-directories are missing (I pruned a few in the snapshot to save space). On the other hand, I do not always use cmake and have not tested it in a while. I guess you were trying to use cmake from the beginning? I'll try to run an install with qe 7.1 and cmake and see what happens. brb

**Reply by paulatz** [reply]
I see what happens, qe-7.1 tries to get the d3q code directly from github, but development has continued and it may not be compatible anymore. I may put a compatible version in the "master" branch, but that would still break for people with different QE versions

**Reply by paulatz** [reply]
I have quickly put together a fix, it works like this:

- you download standard qe 7.1 and configure cmake as usual, with the option -DQE_ENABLE_PLUGINS=d3q (and any other oyu may need)
- grab the updated d3q code https://anharmonic.github.io/release/d3q-cmake.tar.xz
- get rid the directory in external/d3q (which contains the automatically downloaded code) and replace it with the content of d3q-cmake.tar.xz
- go back to your build directory and build. The build process will try to pull a new git version which will give a few errors, but should compile fine nevertheless.

Let me know if it helps, I'll try to set up a more permanent solution.

hth

**Reply by jamie-arcc** [reply]
I was able to get it to compile so now I'm just waiting to see if it works.  I avoided the error messages by commenting out the `qe_git_submodule_update(external/d3q)` in `external/CMakeLists.txt`.  It looks like the QE team just needs to have the correct commit to checkout in their cmake code with the way they're doing it... I feel like they should be testing that before making a release if they're going to have such a feature.

Thanks for your help!

**Reply by paulatz** [reply]
I agree, but it only a team of 1 person taking care of this part. We'll try to get it right for the next release. Thank you for reporting
